### PR TITLE
Fix valid + falsy propvalues

### LINKED ISSUE #562: Attribute selector doesn't work for falsy prop values
For example, 
`wrapper.find('[data-page=0]')` - this selector returns all nodes with _data-page_ attribute, 
but `wrapper.find('[data-page=1]') ` works fine.

Looks like a problem with conditional statement (_instHasProperty_ function)
```javascript
if (propValue) {
    return nodePropValue === propValue;
}
return nodeProps.hasOwnProperty(propKey);
```

### PR DESCRIPTION
**Pull request by ljharb**
Fixes #562.

I also consolidated some code.

### CODE REVIEW

**Comment by nfcampos** [comment]
is `''` a valid prop value as well?

**Comment by ljharb** [comment]
@nfcampos good call, i'll cover that also.

**Comment by nfcampos** [comment]
and while you're at it `null` and `NaN` as well?

**Comment by ljharb** [comment]
Are `null` and `NaN` actually selectable propValues with the string selector syntax? This PR is quickly growing large in scope…

**Comment by ljharb** [comment]
Doesn't React strip out null properties?

**Comment by nfcampos** [comment]
only `undefined` I believe
https://esnextb.in/?gist=899331d29dda61430d88278ddb44a5d8

**Comment by ljharb** [comment]
Looks like it doesn't strip out `false` either.

**Comment by ljharb** [comment]
Ready for a stamp!

File: src/Utils.js [hunk 1]
```diff
@@ -213,6 +213,19 @@ export function coercePropValue(propName, propValue) {
     return propValue;
   }
 
+  // can be the empty string
+  if (propValue === '') {
+    return propValue;
+  }
+
+  if (is('NaN', propValue)) {
```
Comment (nfcampos) [inline src/Utils.js]: any reason this isn't `propValue !== 'NaN'`?

Comment (nfcampos) [inline src/Utils.js]: I mean ===

File: src/Utils.js [hunk 2]
```diff
@@ -213,6 +213,19 @@ export function coercePropValue(propName, propValue) {
     return propValue;
   }
 
+  // can be the empty string
```
Comment (nfcampos) [inline src/Utils.js]: should these three checks here or after the trim like the ones for true and false?

File: src/MountedTraversal.js [hunk 3]
```diff
@@ -14,6 +13,9 @@ import {
   nodeHasType,
 } from './Utils';
 import {
+  nodeHasProperty,
```
Comment (nfcampos) [inline src/MountedTraversal.js]: should this go into Utils?

File: src/Utils.js (hunk 1 above)
Comment (ljharb) [inline src/Utils.js]: i'll change it; i'd been using a literal `NaN` and then updated it to take in a string

File: src/MountedTraversal.js (hunk 3 above)
Comment (ljharb) [inline src/MountedTraversal.js]: Sure, why not

File: src/Utils.js (hunk 2 above)
Comment (ljharb) [inline src/Utils.js]: It didn't feel right having `' '` be treated as the empty string, or `'   NaN'` be treated as a literal NaN - I think the current behavior of throwing for those cases is appropriate.

Comment (nfcampos) [inline src/Utils.js]: makes sense, but shouldn't we do the same for true and false then?

File: test/ShallowTraversal-spec.jsx [hunk 4]
```diff
@@ -112,6 +112,35 @@ describe('ShallowTraversal', () => {
...
+    it('should work with null', () => {
+      expect(nodeHasProperty(<div foo={null} />, 'foo', 'null')).to.equal(true);
+      expect(nodeHasProperty(<div foo={0} />, 'foo', 'null')).to.equal(false);
+    });
+
+    it('should work with false', () => {
+      expect(nodeHasProperty(<div foo={false} />, 'foo', 'false')).to.equal(true);
+      expect(nodeHasProperty(<div foo={0} />, 'foo', 'false')).to.equal(false);
+    });
+
```
Comment (nfcampos) [inline test/ShallowTraversal-spec.jsx]: does this work for `Infinity`?

File: src/Utils.js (hunk 2 above)
Comment (ljharb) [inline src/Utils.js]: Yes, but not in this PR, because that would make it a breaking change

Comment (nfcampos) [inline src/Utils.js]: yep I know, but is it actually worth it to change it seeing as it'd be a such a minor, yet breaking, change?

File: test/ShallowTraversal-spec.jsx (hunk 4 above)
Comment (ljharb) [inline test/ShallowTraversal-spec.jsx]: Yes; added a test to confirm.

**Comment by nfcampos** [comment]
LGTM

**Pull request merged**